src/Network/MQTT/Client.hs

```diff
@@ -0,0 +1,115 @@
+                                               , writeTQueue )
+import           Control.Concurrent.STM.TVar   ( TVar, newTVar, writeTVar )
+import           Control.Monad                 ( forM_, when )
+import           Control.Monad.STM             ( STM, atomically )
+import           Control.Concurrent            ( forkIO )
+
+type SocketHandle = (Socket, Handle)
+
+data Config = Config { hostName :: HostName
+                     , port     :: String
+                     }
+
+data ClientState =
+  ClientState
+  { -- | All packets from the this queue will be sent to the Server
+    sendQueue        :: TQueue Packet
+    -- | Whether the Client is connected to the Server
+  , connected        :: TVar Bool
+    -- | Whether the Client received Pingresp packet from the Server
+  , pingRespReceived :: TVar Bool
+  }
+  deriving (Eq)
+
+initState :: STM ClientState
+initState = do
+  sendQueue <- newTQueue
+  conn <- newTVar False
+  pingRespReceived <- newTVar False
+  return $ ClientState
+             sendQueue
+             conn
+             pingRespReceived
+
+conn :: HostName -> String -> IO SocketHandle
+conn hostname port = do
+    addrinfos <- getAddrInfo Nothing (Just hostname) (Just port)
+    let serveraddr = head addrinfos
+    sock <- socket (addrFamily serveraddr) Stream defaultProtocol
+    setSocketOption sock KeepAlive 1
+    connect sock (addrAddress serveraddr)
+    h <- socketToHandle sock ReadWriteMode
+    hSetBuffering h (BlockBuffering Nothing)
+    hSetBinaryMode h True
+    return (sock, h)
+
+sendPacket :: SocketHandle -> Packet -> IO ()
+sendPacket (_, handle) packet = do
+    hPutBuilder handle (encodePacket packet)
+    hFlush handle
+
+closeConn :: SocketHandle -> IO ()
+closeConn (_, handle) = print "Connection closed" >> hClose handle
+
+disconn :: SocketHandle -> IO ()
+disconn socketHandle = do
+  sendPacket socketHandle (DISCONNECT DisconnectPacket)
+  closeConn socketHandle
+
+------------------------------------------------------------------------------------------------
+
+sender' :: ClientState -> SocketHandle -> IO ()
+sender' state socketHandle = do
+  packet <- atomically $ readTQueue (sendQueue state)
+  sendPacket socketHandle packet
+  print $ "send: " ++ show packet
+  sender' state socketHandle
+
+receiver' :: ClientState -> SocketHandle -> IO ()
+receiver' state@ClientState{..} socketHandle = do
+  (_, pkgs) <- parseMany (fst socketHandle) (parse parsePacket) (parse parsePacket)
+  forM_ pkgs $ \p -> do
+    print $ "receive: " ++ show p
+    case p of
+      (CONNACK ConnackPacket{..}) -> do
+        when (connackReturnCode /= Accepted) (disconn socketHandle)
+        atomically $ writeTVar connected True
+      (PINGRESP PingrespPacket) -> atomically $ writeTVar pingRespReceived True
+      _ -> return ()
+  receiver' state socketHandle
+
+sender :: ClientState -> Packet -> IO ()
+sender ClientState{..} p = atomically $ do
```

Comment:
I would create `sendPingreq` method, as that will be really hard to extend this one to other types of packets.